### ISSUE TITLE
Make redis prefix configurable

### ISSUE THREAD

**Issue by egze** [tompave/fun_with_flags#36]
Would be super nice to have the prefix for keys configurable. I can make a PR.

**Comment by tompave** [comment]
Hi, thanks for the suggestion.
So that I can better understand the requirement, what's your use case?

**Comment by egze** [comment]
I'm building a microservice with REST around FunWithFlags and our DevOps want to have Redis keys prefixed with the service names. It's a convention across all services in the company.

**Comment by tompave** [comment]
I see. Are you using a single Redis instance for all services?

It sounds like the kind of thing that could be solved with either:
* Don't share the redis instance, use one per service.
* Use the ecto persistence adapter, so that you can keep the flags in the DB.

I'm suggesting alternatives because I'd rather not overcomplicate the configuration options if I can avoid it. In my opinion sharing a datastore among services is not the greatest thing (although I understand that it could be required because of costs), so this addition seems only useful to deal with a system design that I would not recommend anyway.

**Comment by egze** [comment]
Sure, there are ways around it, but then it often means going against internal policies or changing existing infrastructure. A config for the prefix sounds like a simple change and would not make the code more complex. But would mean a lot for devs like me that have certaing infrastructure restrictions.

Also like I said - I would more than happy to make a PR.

**Comment by tompave** [comment]
Yes, I agree that it would be a simple change, but I don't think that it's a change that I'd want to include and maintain. This especially because once it's merged and it's part of the public API, it's there to stay:

> Rule # 1 of open-source: no is temporary, yes is forever.
> [Solomon Hykes, 30 Mar 2016](https://twitter.com/solomonstre/status/715277134978113536)

With regard to going against internal policies and infrastructure, I understand why this is causing you issues right now. However, I wouldn't add and then maintain support for a use case that I think is flawed.

You're not stuck with my opinion though, and if you disagree you can add your own persistence adapter! 🙂 

1. Add this package from hex.pm as it is.
2. Add a file to your project, copying the source from [`FunWithFlags.Store.Persistent.Redis`](https://github.com/tompave/fun_with_flags/blob/v1.2.1/lib/fun_with_flags/store/persistent/redis.ex). Give it a different name, e.g. `MyApp.CustomFwfRedisPersistence`.
3. Modify it to do what you need. The important things is that it needs to implement these functions: `worker_spec/0`, `get/1`, `put/2`, `delete/2`, `delete/1`, `all_flags/0`, `all_flag_names/0`. Now that I think about it, I suppose I should define a `Behaviour` to make this clearer.
4. Configure the package to use your adapter, with:
    ```elixir
    config :fun_with_flags, :persistence, adapter: MyApp.CustomFwfRedisPersistence
    ```

That should allow you to set the Redis keys in any way you want. Please give it a try and let me know if it doesn't work!

**Comment by egze** [comment]
Aha, this sounds like a nice idea. Thanks, I'll give it a try!

**Comment by tompave** [comment]
Closing this as I expect that it's been resolved.

**Comment by chess4ever** [comment]
Hello @tompave , I would like to reopen this issue. Actually with redis6 ACLs it is normal to have prefixes for each application in a shared redis instance, so I think having the prefix configurable is a must even for basic use cases.
Having a custom adapter should be just for more advanced and specific needs in my opinion.

**Comment by tompave** [comment]
Hey there. The answer is still no, sorry.

I still believe in the value of simplicity and avoiding too many options, and I don't think that Redis prefixes should be supported in the default persistence adapter.

Providing a custom Redis persistence adapter is the recommended way to support this use case, and you can even release it on Hex.pm if you want.

For context, this issue is from March 2019, and in May 2019 I released [v1.4.0](https://github.com/tompave/fun_with_flags/blob/master/CHANGELOG.md#v140) of the package, which provides a documented Elixir Behaviour to implement custom adapters.
